Fixes #37102 - webpack 5

MariaAga opened the pull request:
Details and discussion here: https://community.theforeman.org/t/resurrection-of-the-client-side-infrastructure-upgrade-effort/32340

ehelms commented on bundler.d/assets.rb:
```diff
@@ -4,7 +4,7 @@
   gem 'po_to_json', '~> 1.1'
   gem 'execjs', '>= 1.4.0', '< 3.0'
-  gem 'uglifier', '>= 1.0.3'
+  gem "terser", "~> 1.1"
```
Can the change from uglifier to terser be done ahead of time with our webpack 3 setup?

MariaAga replied:
Yeah, we should, thanks

ehelms replied:
@MariaAga if you'll open a PR with this change, I'll review it and test the packaging.

ehelms replied:
I started the packaging for this https://github.com/theforeman/foreman-packaging/pull/9877 once you pull it out to it's own PR.

ehelms replied:
I've opened a PR for this change -- https://github.com/theforeman/foreman/pull/9913

ehelms commented on package.json:
```diff
@@ -20,11 +20,14 @@
-    "react-intl": "^2.8.0"
+    "os-browserify": "^0.3.0",
+    "react-intl": "^2.8.0",
+    "stream-browserify": "^3.0.0"
```
Where is this used?

MariaAga replied:
Leftovers, removed. thanks

ehelms commented on package.json:
```diff
@@ -38,38 +41,38 @@
-    "node-sass": "^4.5.0",
-    "optimize-css-assets-webpack-plugin": "^3.2.0",
+    "mini-css-extract-plugin": "2.7.5",
+    "node-sass": "6.0.0",
```
Is this an update we can do independent of the webpack change?

MariaAga replied:
no - `sass-loader` only supports `node-sass` v6 from `sass-loader` v11.1.0
and `sass-loader` v11.0.0 only supports wepack 5

ehelms commented on package.json:
```diff
@@ -38,38 +41,38 @@
-    "compression-webpack-plugin": "~1.1.11",
     "cross-env": "^5.2.0",
-    "css-loader": "^0.23.1",
+    "css-loader": "4.3.0",
```
Is this an update we can do independent of the webpack change?

MariaAga replied:
`css-loader` 4 requires webpack 4, so we can't do it separetly

ehelms commented:
I've tried to pick out a few spots where my goal is to identify any changes we can make independently so that we are able to do the small update round trip with packaging and then reduce the scope of changes required at one time for this PR.

ekohl commented on developer_docs/getting-started.asciidoc:
````diff
@@ -36,10 +36,8 @@ Following steps are required to setup a webpack development environment:
      ```bash
-     ./node_modules/.bin/webpack-dev-server \
+      npx webpack \
        --config config/webpack.config.js \
````
```suggestion
       --config config/webpack.config.js
```

ekohl commented on app/assets/javascripts/late_load.js:
```diff
@@ -0,0 +1,34 @@
+    const loadJS = new Event('loadJS');
+    document.dispatchEvent(loadJS);
+  });
+}
```
Line endings please

ekohl commented on app/helpers/reactjs_helper.rb:
```diff
@@ -11,58 +10,60 @@ def react_component(name, props = {})
-    global_css_tags(global_plugins_list).join.html_safe
+  def js_tags_for(requested_plugins)
+    requested_plugins.map do |plugin|
+      javascript_tag("window.tfm.tools.loadPluginModule('/webpack/#{plugin.to_s.tr('-', '_')}','#{plugin.to_s.tr('-', '_')}','./index');".html_safe)
```
In `global_js_tags` you use `plugin[:id]` and here just `plugin`. That should be consistent.

ekohl commented on app/helpers/reactjs_helper.rb:
```diff
@@ -11,58 +10,60 @@ def react_component(name, props = {})
-  def webpacked_plugins_with_global_js
-    global_js_tags(global_plugins_list).join.html_safe
+  def other_webpack_plugin(plugin_name, file)
+    javascript_tag("window.tfm.tools.loadPluginModule('/webpack/#{plugin_name.to_s.tr('-', '_')}','#{plugin_name.to_s.tr('-', '_')}','./#{file}_index');".html_safe)
```
Given the repetition of this conversion I think we should have a common way for it. I'm not exactly sure where that should live, but the Foreman plugin code feels like a natural place. Perhaps we already have such a thing.

ekohl commented on app/helpers/reactjs_helper.rb:
```diff
@@ -11,58 +10,60 @@ def react_component(name, props = {})
-      end
+  def get_webpack_foreman_vendor_js
+    if ENV['RAILS_ENV'] == 'production'
+      javascript_include_tag("/webpack/#{File.basename(Dir.glob('public/webpack/foreman-vendor*production*js')[0])}")
```
Please avoid globbing for these methods. If there's some left over file it may end up using the wrong file. It also creates additional IO for every request at runtime. Previously we had some JSON file which had the mapping. Isn't that something that can still be done?

ShimShtein replied:
Previously we had to create the JSON manifest files to make rails helpers to work correctly (it's a hack built upon the `webpack-rails` gem). Now that we don't use the gem and don't reference the files from Ruby code, I don't see a good reason to keep that special configuration in webpack to generate the manifest file. 
This concept inspired heavily b jsbundling-rails gem, which is the default option for rails7. The idea behind it is to reduce knowledge of the JS building process and serve the files as pure static files. 

In our specific case, there always should be only one vendor file in the folder, but we don't know the hash ID for it. Would it be better to assert single file here, and prevent the app from loading in case more than one file exists?

ekohl replied:
Globbing can be unpredictable. sprockets also has a random component in its manifest filename and they at least have a warning if multiple files are found:

https://github.com/rails/sprockets/blob/5b040f3bec503ded8a98c0c911c7a77dd1f5bf1b/lib/sprockets/manifest_utils.rb#L40-L43

And you still haven't addressed my concern that this isn't stored in the application program, but loaded every time. I don't insist on a manifest, but your current approach causes additional IO and can make upgrades less predictable (because there can be a point where the new files are deployed, but the application not yet restarted).

I'm not sure what the best place is, but an application initializer would be my first thought.

ShimShtein replied:
> And you still haven't addressed my concern that this isn't stored in the application program, but loaded every time. I don't insist on a manifest, but your current approach causes additional IO and can make upgrades less predictable (because there can be a point where the new files are deployed, but the application not yet restarted).

What about having the filename cached once?
Something along the lines of:
``` ruby  
  def self.manifest_file
    logger.warn("too many foreman-vendor files") if Dir.glob('public/webpack/foreman-vendor*production*js').length > 1
     @manifest_file ||= "/webpack/#{File.basename(Dir.glob('public/webpack/foreman-vendor*production*js')[0])}"
  end

  def get_webpack_foreman_vendor_js
      javascript_include_tag(self.manifest_file)
  end
```

This would make the upgrade predictable (until the next restart) and avoid extra IO.

ekohl replied:
I would be worried about multiple workers. When Puma forks of a new worker, would it determine the manifest file again? In production mode I would expect it to be part of an initializer so it happens before forking.

And a small nit that I would use:
```ruby
def self.manifest_file
  @manifest_file ||= begin
    files = Dir.glob('public/webpack/foreman-vendor*production*js')
    logger.warn("too many foreman-vendor files") if files.length > 1
    "/webpack/#{File.basename(files.first)}"
  end
end
```

Because your logger statement would still glob on every request.

MariaAga replied:
I was thinking of doing:
``` ruby
  def read_webpack_manifest
    root = File.expand_path(File.dirname(__FILE__) + "/../..")
    file = File.read(root + '/public/webpack/manifest.json')
    JSON.parse(file)
  end

  def get_webpack_foreman_vendor_js
    data = read_webpack_manifest
    foreman_vendor_js = data['assetsByChunkName']['foreman-vendor'].find { |value| value.end_with?('.js') }
    javascript_include_tag("/webpack/#{foreman_vendor_js}")
  end
```

ShimShtein replied:
@MariaAga  The whole idea is to get rid of the manifest, but if we decide to leave the manifest we shall use it of course.

ekohl commented on app/helpers/reactjs_helper.rb:
```diff
@@ -11,58 +10,60 @@ def react_component(name, props = {})
-        javascript_include_tag(*webpack_asset_paths("#{plugin[:id]}:#{file}", :extension => 'js'))
-      end
+  def get_webpack_foreman_vendor_js
+    if ENV['RAILS_ENV'] == 'production'
```
```suggestion
    if Rails.env.production?
```

ekohl commented on app/helpers/reactjs_helper.rb:
```diff
@@ -11,58 +10,60 @@ def react_component(name, props = {})
+    if ENV['RAILS_ENV'] == 'production'
+      javascript_include_tag("/webpack/#{File.basename(Dir.glob('public/webpack/foreman-vendor*production*js')[0])}")
+    else
+      javascript_include_tag("/webpack/#{File.basename(Dir.glob('public/webpack/foreman-vendor*development*js')[0])}")
```
This implies in testing we also use the development version. Is that correct?

MariaAga replied:
The filename depends on `NODE_ENV` should I change `RAILS_ENV`  to `NODE_ENV`?
can `NODE_ENV` be something that is not `production` or `development`?

ekohl replied:
Not sure about `NODE_ENV`. Like with https://github.com/theforeman/foreman/pull/9834#discussion_r1329837249, can't we write some JSON file at a predictable location with the filename? I suspect that determining this at runtime is likely to cause problems at some point.

ShimShtein replied:
As far as tests go we don't need webpack file to be loaded anyway (we don't do integration tests using `RAILS_ENV=test`). 

As I have mentioned before: since we need a single file there anyway, would it be OK to assert and prevent loading the app in case more than one file exists? Having the manifest file requires knowledge transfer between the webpack and rails parts of the code, and I try to simplify both webpack configuration and rails loading process.

In a long term, I think this part also should be part of the foreman federated module Maria already mentioned in the thread. This should remove the need to know the exact filename entirely.

ekohl replied:
Shouldn't it be:
```suggestion
      javascript_include_tag("/webpack/#{File.basename(Dir.glob('public/webpack/foreman-vendor*#{ENV.fetch('NODE_ENV', 'production')}*js')[0])}")
```

ehelms commented:
Should I be able to npm install at this point? I am seeing the following:

```
$ npm install
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! While resolving: TheForemanDevDeps@3.9.0
npm ERR! Found: webpack@5.75.0
npm ERR! node_modules/webpack
npm ERR!   dev webpack@"5.75.0" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer webpack@"^4.0.0" from optimize-css-assets-webpack-plugin@4.0.3
npm ERR! node_modules/optimize-css-assets-webpack-plugin
npm ERR!   dev optimize-css-assets-webpack-plugin@"^4.0.0" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
```

adamruzicka commented:
@ehelms Seems to work for me. Is your katello patched as well? Did you wipe node_modules before npm install? What version of node do you have?

ehelms commented:
> @ehelms Seems to work for me. Is your katello patched as well? Did you wipe node_modules before npm install? What version of node do you have?

No Katello involved in my test. I tried this with 18 and 16 as that is what I had local. Perhaps I made a wrong assumption that this would already enable doing it with newer Nodes. I can re-test with 12.

adamruzicka commented:
I have node 14, but I can't tell if that has any impact

MariaAga commented:
Node 14 is recommended and tested. Making node 16+ available will be possible after this PR but requires some changes  that should be done it a new PR.
Also I recommend deleting `package-lock.json` and `node_modules` when testing this PR.

ehelms commented:
> Node 14 is recommended and tested. Making node 16+ available will be possible after this PR but requires some changes that should be done it a new PR.

Looks like this also currently requires Node 14?

```
npm ERR! code ENOTSUP
npm ERR! notsup Unsupported engine for css-minimizer-webpack-plugin@4.2.2: wanted: {"node":">= 14.15.0"} (current: {"node":"12.20.1","npm":"6.14.10"})
npm ERR! notsup Not compatible with your version of node/npm: css-minimizer-webpack-plugin@4.2.2
npm ERR! notsup Not compatible with your version of node/npm: css-minimizer-webpack-plugin@4.2.2
npm ERR! notsup Required: {"node":">= 14.15.0"}
npm ERR! notsup Actual:   {"npm":"6.14.10","node":"12.20.1"}
```

If yes, and there is no way to support Node 12, we will have to consider that transition in our build environment. It would be nice to be able to only do one change at a time. We need to understand if that is possible or impossible.

ShimShtein commented:
> npm ERR! notsup Unsupported engine for css-minimizer-webpack-plugin@4.2.2: wanted: {"node":">= 14.15.0"} (current: {"node":"12.20.1","npm":"6.14.10"})

@MariaAga , can we use `css-minimizer-webpack-plugin@3.4.1` ? I see they dropped the support for node 12 in >=4.0.0: https://github.com/webpack-contrib/css-minimizer-webpack-plugin/commit/88a5674f4e20dcdf6f73f6406aea9749a3d22a17

@ehelms , can you try locally if you can work with lower version of the plugin?

ehelms commented:
> > npm ERR! notsup Unsupported engine for css-minimizer-webpack-plugin@4.2.2: wanted: {"node":">= 14.15.0"} (current: {"node":"12.20.1","npm":"6.14.10"})
> 
> @MariaAga , can we use `css-minimizer-webpack-plugin@3.4.1` ? I see they dropped the support for node 12 in >=4.0.0: [webpack-contrib/css-minimizer-webpack-plugin@88a5674](https://github.com/webpack-contrib/css-minimizer-webpack-plugin/commit/88a5674f4e20dcdf6f73f6406aea9749a3d22a17)
> 
> @ehelms , can you try locally if you can work with lower version of the plugin?

Looks like it's a no go, but due to webpack itself:

```
pm ERR! notsup Unsupported engine for webpack-cli@5.0.1: wanted: {"node":">=14.15.0"} (current: {"node":"12.20.1","npm":"6.14.10"})
npm ERR! notsup Not compatible with your version of node/npm: webpack-cli@5.0.1
npm ERR! notsup Not compatible with your version of node/npm: webpack-cli@5.0.1
npm ERR! notsup Required: {"node":">=14.15.0"}
npm ERR! notsup Actual:   {"npm":"6.14.10","node":"12.20.1"}
```

ehelms commented:
Based on the nodejs 14 requirement, I was curious to, from a build perspective, look at if our packages would build against 14 and then if Foreman would. I haven't tested the results, but everything built (and thanks to Copr) I have a repo with only those nodejs and Foreman package in it:

https://copr.fedorainfracloud.org/coprs/ehelms/foreman-nightly-staging/

ehelms commented:
> Based on the nodejs 14 requirement, I was curious to, from a build perspective, look at if our packages would build against 14 and then if Foreman would. I haven't tested the results, but everything built (and thanks to Copr) I have a repo with only those nodejs and Foreman package in it:
> 
> https://copr.fedorainfracloud.org/coprs/ehelms/foreman-nightly-staging/

I took the Foreman build here for a test drive and poked around at the UI -- I found no issues. This leads me to believe we can do an update to Node 14 in our build environment with our current code base prior to this change and reduce the friction.

Looking back at history, [1] was the reason we had ended up in a chicken-egg problem with prior attempts because Node 14 in the RPM environment came with an older NPM. However, I believe that we can update to Node 14 as is cleanly and that [1] is not as relevant to get to Webpack 5 (and can be a nice to have later on).

@MariaAga @ShimShtein can you confirm?

[1] https://github.com/theforeman/foreman-js/pull/404

ehelms commented:
https://github.com/theforeman/tool_belt/pull/569 -- this will swap build environment to Node 14.

ehelms commented:
Can you rebase this? I've done a round of updates to Node 14.

ekohl commented on app/helpers/layout_helper.rb:
```diff
@@ -98,6 +98,28 @@ def javascript(*args)
+
+  def webpack_asset_paths(plugin_name, extension: 'js')
+    if extension == 'js'
+      Foreman::Deprecation.deprecation_warning('3.9', '`webpack_asset_paths` is deprecated, use `content_for(:javascripts) { webpacked_plugins_js_for(plugin_name) }` instead.')
```
The version passed in is the one where it gets removed. 3.9 implies it should be removed immediately since that's the current nightly.

MariaAga commented:
Thanks everyone for your input, Updated some loading order logic so Jquery should work as expected now. The PR still needs some cleaning up + info comments + test fixes

ekohl commented on Dockerfile:
```diff
@@ -1,7 +1,7 @@
 FROM quay.io/centos/centos:stream8 as base
 ARG RUBY_VERSION="2.7"
-ARG NODEJS_VERSION="12"
+ARG NODEJS_VERSION="14"
```
Something that we should have already done in develop.

ekohl replied:
https://github.com/theforeman/foreman/pull/9916

ekohl commented on app/assets/javascripts/host_edit_interfaces.js:
```diff
@@ -413,9 +413,9 @@ $(document).on('change', '.interface_mac', function(event) {
-    var interface = $('#interfaceModal').find('.interface_mac');
-    var mac = interface.val();
-    var baseurl = interface.attr('data-url');
+    var interface_ = $('#interfaceModal').find('.interface_mac');
```
For my own understanding: why is this needed? Is `interface` now a reserved value or a global variable?

MariaAga replied:
It is, and it was a small enough change that I didnt test excatly why the js didnt like us using `interface` . https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode#future-proofing_javascript

ekohl replied:
Ah ok. That explains it and IMHO that makes it a candidate to already merge it outside of this PR to make it smaller.

MariaAga replied:
https://github.com/theforeman/foreman/pull/9937

ekohl commented on app/helpers/layout_helper.rb:
```diff
@@ -98,6 +98,28 @@ def javascript(*args)
   end
 
+  def javascript_include_tag(*params, **kwargs)
+    # Workaround for overriding javasctipt load with webpack_asset_paths, should be removed when webpack_asset_paths is removed
```
```suggestion
    # Workaround for overriding javascript load with webpack_asset_paths, should be removed when webpack_asset_paths is removed
```

ekohl commented on app/helpers/layout_helper.rb:
```diff
@@ -98,6 +98,28 @@ def javascript(*args)
 
+  def javascript_include_tag(*params, **kwargs)
+    # Workaround for overriding javasctipt load with webpack_asset_paths, should be removed when webpack_asset_paths is removed
+    if kwargs.is_a?(Hash) && kwargs[:source] == "webpack_asset_paths"
```
kwargs should always be a `Hash`
```suggestion
    if kwargs[:source] == "webpack_asset_paths"
```

ekohl commented on app/helpers/layout_helper.rb:
```diff
@@ -98,6 +98,28 @@ def javascript(*args)
+    end
+  end
+
+  def webpack_asset_paths(plugin_name, extension: 'js')
```
I was surprised to see why a new function showed up with just deprecations. Perhaps suggest where it came from?
```suggestion
  # @deprecated Previously provided by webpack-rails
  def webpack_asset_paths(plugin_name, extension: 'js')
```

ekohl commented on app/helpers/reactjs_helper.rb:
```diff
@@ -24,7 +21,25 @@ def webpacked_plugins_with_global_js
+  end
+
+  def read_webpack_manifest
+    root = File.expand_path(File.dirname(__FILE__) + "/../..")
+    file = File.read(root + '/public/webpack/manifest.json')
+    JSON.parse(file)
```
There's `Rails.root` (instance of `Pathname`) as a short hand so it can be simplified to:
```suggestion
    JSON.parse(Rails.root.join('public/webpack/manifest.json').read)
```

ekohl commented on app/helpers/reactjs_helper.rb:
```diff
@@ -24,7 +21,25 @@ def webpacked_plugins_with_global_js
 
   def webpacked_plugins_css_for(*plugin_names)
-    css_tags_for(select_requested_plugins(plugin_names)).join.html_safe
+    Foreman::Deprecation.deprecation_warning('3.9', '`webpacked_plugins_css_for` is deprecated, plugin css is already loaded.')
```
```suggestion
    Foreman::Deprecation.deprecation_warning('3.12', '`webpacked_plugins_css_for` is deprecated, plugin css is already loaded.')
```

ekohl commented on app/helpers/reactjs_helper.rb:
```diff
@@ -24,7 +21,25 @@ def webpacked_plugins_with_global_js
+  end
+
+  def get_webpack_foreman_vendor_js
+    data = read_webpack_manifest
+    foreman_vendor_js = data['assetsByChunkName']['foreman-vendor'].find { |value| value.end_with?('.js') }
```
This still reads the manifest on every request, possibly multiple times. Would it make sense to cache this?

MariaAga replied:
@ekohl Yeah, for how long?

ekohl commented on app/helpers/reactjs_helper.rb:
```diff
@@ -39,30 +54,21 @@ def select_requested_plugins(plugin_names)
   def js_tags_for(requested_plugins)
     requested_plugins.map do |plugin|
-      javascript_include_tag(*webpack_asset_paths(plugin.to_s, :extension => 'js'))
+      javascript_tag("window.tfm.tools.loadPluginModule('/webpack/#{plugin.to_s.tr('-', '_')}','#{plugin.to_s.tr('-', '_')}','./index');".html_safe)
```
I had a hard time reading this. How about:
```suggestion
      name = plugin.to_s.tr('-', '_')
      javascript_tag("window.tfm.tools.loadPluginModule('/webpack/#{name}', '#{name}', './index');".html_safe)
```

ekohl commented on app/helpers/reactjs_helper.rb:
```diff
@@ -39,30 +54,21 @@ def select_requested_plugins(plugin_names)
-    requested_plugins.map do |plugin|
-      plugin[:files].map do |file|
-        stylesheet_link_tag(*webpack_asset_paths("#{plugin[:id]}:#{file}", :extension => 'css'))
+        javascript_tag("window.tfm.tools.loadPluginModule('/webpack/#{plugin[:id].to_s.tr('-', '_')}','#{plugin[:id].to_s.tr('-', '_')}','./#{file}_index');".html_safe)
```
Similar as above:
```suggestion
        name = plugin[:id].to_s.tr('-', '_')
        javascript_tag("window.tfm.tools.loadPluginModule('/webpack/#{name}', '#{name}', './#{file}_index');".html_safe)
```

ekohl commented on app/helpers/reactjs_helper.rb:
```diff
@@ -39,30 +54,21 @@ def select_requested_plugins(plugin_names)
-    requested_plugins.map do |plugin|
-      stylesheet_link_tag(*webpack_asset_paths(plugin.to_s, :extension => 'css'))
-    end
+    Foreman::Deprecation.deprecation_warning('3.12', '`css_tags_for` is deprecated, No need to load CSS separately, since it should be referneced from the corresponding JS file.')
```
```suggestion
    Foreman::Deprecation.deprecation_warning('3.12', '`css_tags_for` is deprecated, No need to load CSS separately, since it should be referenced from the corresponding JS file.')
```

ekohl commented on package.json:
```diff
@@ -5,7 +5,7 @@
   "private": true,
   "engines": {
-    "node": "<16.0.0"
+    "node": ">12.0.0 <16.0.0"
```
I think it's safe to raise this to 14 by now
```suggestion
    "node": ">14.0.0 <16.0.0"
```

ekohl commented on package-exclude.json:
```diff
@@ -30,8 +30,7 @@
         "webpack-bundle-analyzer",
-        "webpack-dev-server",
-        "webpack-dev-server-without-h2",
+        "webpack-stats-plugin",
```
The webpack config has this line:
```js
var StatsWriterPlugin = require('webpack-stats-plugin').StatsWriterPlugin;
```

That makes me think webpack will crash/fail if the plugin isn't present, but we do need to run webpack in a packaged environment meaning we can't skip it. I didn't test this though and perhaps JS will happily proceed if the package can't be loaded, but it'd surprise me.

MariaAga replied:
Removed `webpack-stats-plugin` from the exclude list as it creates the manifest file which ruby reads to know what to load.

ekohl commented on webpack/assets/javascripts/react_app/common/variables.scss:
```diff
@@ -1,3 +1,3 @@
-@import '~@theforeman/vendor/scss/variables';
+@import 'node_modules/@theforeman/vendor/scss/variables.scss';
```
Is it normal to include `node_modules` in a path?

MariaAga replied:
jenkins doesnt understand `~@theforeman` for some reasons and I'm trying to see whats going on there (which is also why there are so many pushes today).
Locally this works fine

ekohl replied:
Sounds like something that needs further investigation. I bet it's something subtle in some version difference of some library or something being present/absent.

ekohl commented on webpack/assets/javascripts/react_app/common/globalIdHelpers.js:
```diff
@@ -7,6 +7,8 @@
  * together into a single string and encoded as base64.
  */
 
+import { Buffer } from 'buffer';
```
Looks like something we can already merge in a preparation PR, correct?

MariaAga replied:
https://github.com/theforeman/foreman/pull/9909

ekohl commented on webpack/assets/javascripts/react_app/components/common/DateTimePicker/DateTimePicker.js:
```diff
@@ -93,7 +93,7 @@ class DateTimePicker extends React.Component {
             position={placement}
             className="container"
-            style={{ minWidth: '50em', width: '33%'}}
+            style={{ minWidth: '50em', width: '33%' }}
```
Something we can already merge, right?

MariaAga replied:
https://github.com/theforeman/foreman/pull/9910

ekohl commented on app/views/users/_form.html.erb:
```diff
@@ -140,9 +140,11 @@
 
 <% if @user.cached_usergroups.any? %>
-  <script>
+  <% content_for(:javascripts) do %>
```
Is this a change we could already merge? I saw a few more `content_for` changes like this and wonder if we can trim down the size of this PR a bit by doing more preparations.

MariaAga replied:
https://github.com/theforeman/foreman/pull/9936

ekohl reviewed (commented):
I haven't done any testing, but I've tried to identify areas we can already merge to current develop and things that I think should be addressed. Also had a few items that I'd like to better understand.

MariaAga commented:
Tests are failing because production is not being built right

ekohl commented on app/helpers/layout_helper.rb:
```diff
@@ -98,6 +98,30 @@ def javascript(*args)
+
+
+  def webpack_asset_paths(plugin_name, extension: 'js')
+    # @deprecated Previously provided by webpack-rails
```
I meant to put it above the method. We don't use [yardoc](https://yardoc.org/) that much, but it's a good standard.

MariaAga commented:
Simplified the code and added comments.

ekohl commented on lib/tasks/plugin_assets.rake:
```diff
@@ -43,17 +45,28 @@ task 'plugin:assets:precompile', [:plugin] => [:environment] do |t, args|
-        webpack_bin = ::Rails.root.join('node_modules/webpack/bin/webpack.js')
-        config_file = ::Rails.root.join(::Rails.configuration.webpack.config_file)
-        sh "node --max_old_space_size=2048 #{webpack_bin} --config #{config_file} --bail --env.pluginName=#{@plugin.id}"
+        # TODO: delete this as all plugins get compiled in the regura
```
How would we resolve this?

MariaAga replied:
Updated the code to continue working as before and only compiling one plugin for this task

MariaAga commented:
@evgeni Whats needed for rpm-build not to fail?

evgeni commented:
A new PR, due to a packit bug (https://github.com/packit/packit-service/issues/2262)
But only if you actually want a working rpm build, it's not necessary for merging.

Edit: but even then, I am not sure the build will work given the changes to package json

MariaAga commented:
@evgeni, refined question: How to make the build succeed with the pakcage.json changes?

evgeni commented:
That's something that our setup doesn't support well today.

The TL;DR version is:
- you need an rpm repo with the new dependencies packaged
- you need an updated foreman.spec that reflects those dependency changes

You could probably get this by creating packaging PRs for the deps (that gets you multiple repos, one per nee dep) and a branch that has foreman.spec updated. Then you edit packit.yml to use those and it'll build.

Being able to do such things has been requested a few times recently, so I'll think how we could make this experience better.

evgeni commented on package.json:
```diff
@@ -38,38 +40,34 @@
     "cross-env": "^5.2.0",
-    "css-loader": "^0.23.1",
-    "cssnano": "^4.1.10",
+    "css-loader": "6.8.1",
```
can we use `^6.8.1` here? pinning one exact version seems odd

evgeni commented on package.json:
```diff
@@ -38,38 +40,34 @@
     "highlight.js": "~9.14.0",
-    "node-sass": "^4.5.0",
-    "optimize-css-assets-webpack-plugin": "^3.2.0",
+    "node-sass": "8.0.0",
```
same, this seems too strict (even when today there is only 8.0.0 and 9.0.0)

evgeni commented on package.json:
```diff
@@ -38,38 +40,34 @@
-    "sass-loader": "~6.0.6",
-    "style-loader": "^0.13.1",
+    "sass": "~1.60.0",
+    "sass-loader": "10.2.0",
```
same ;)

evgeni commented on package.json:
```diff
@@ -38,38 +40,34 @@
-    "webpack-stats-plugin": "^0.1.5",
     "tabbable": "~5.2.0",
-    "sass": "~1.60.0"
+    "url-loader": "4.1.1",
+    "webpack": "5.75.0",
```
same :)

evgeni commented on package.json:
```diff
@@ -38,38 +40,34 @@
+    "url-loader": "4.1.1",
+    "webpack": "5.75.0",
+    "webpack-bundle-analyzer": "^4.5.0",
+    "webpack-cli": "5.0.1",
```
what is this used for? I didn't see any `webpack-cli` calls in your PR, or does this also provide the `webpack` binary these days?
(also, too strict version req ;) )

MariaAga replied:
Since were using `npx webpack`, we need to have `webpack-cli`.
from https://webpack.js.org/api/cli :

> Warning
If you want to run webpack using npx please make sure you have webpack-cli installed.

evgeni replied:
TIL, thanks!

evgeni commented:
(rebased to resolve the conflict as the dedicated terser PR used different quotes than this one)

evgeni commented:
[test katello]

jeremylenz commented:
```
WARNING in asset size limit: The following asset(s) exceed the recommended size limit (244 KiB).
This can impact web performance.
Assets: 
  471.js (5.52 MiB)
  860.js (5.75 MiB)
  foreman-vendor.bundle-v12.2.0-production-871f6b5b1ae976d744a2.css.gz (299 KiB)
  images/pfbg_2000.bundle-v12.2.0-production-3b693db01ef2135e3f14fd3604f355aa.jpg (270 KiB)
  images/pfbg_768@2x.bundle-v12.2.0-production-eea28e73c0df6dbbb1ff0ffde38c6c0e.jpg (354 KiB)
  images/pfbg_992@2x.bundle-v12.2.0-production-30e8d31f821a144fe52cf00417648038.jpg (575 KiB)
  foreman-vendor.bundle-v12.2.0-production-871f6b5b1ae976d744a2.css (3 MiB)
  foreman-vendor.bundle-v12.2.0-production-871f6b5b1ae976d744a2.js.gz (2.63 MiB)
  foreman-vendor.bundle-v12.2.0-production-871f6b5b1ae976d744a2.js.map.gz (6.4 MiB)
  foreman-vendor.bundle-v12.2.0-production-871f6b5b1ae976d744a2.js (11.5 MiB)
  foreman-vendor.bundle-v12.2.0-production-871f6b5b1ae976d744a2.js.map (30.5 MiB)

webpack 5.75.0 compiled with 1 warning in 134740 ms
```

Clicked around and everything seems to work fine in the UI!

Checked React, old-school React, AngularJS, and ERB pages in Foreman and Katello.

Too bad about that 134 second build time, tho..

let me know if there's anything else you want me to review :+1:

jeremylenz commented:
By the way, I see a lot of deprecation warnings like this. Are those being fixed?

```
20:49:22 webpack.1 | DEPRECATION WARNING: Using / for division outside of calc() is deprecated and will be removed in Dart Sass 2.0.0.
20:49:22 webpack.1 | 
20:49:22 webpack.1 | Recommendation: math.div($i, 12) or calc($i / 12)
20:49:22 webpack.1 | 
20:49:22 webpack.1 | More info and automated migrator: https://sass-lang.com/d/slash-div
20:49:22 webpack.1 | 
20:49:22 webpack.1 |    ╷
20:49:22 webpack.1 | 17 │         width: #{percentage($i / 12) - 1};
20:49:22 webpack.1 |    │                             ^^^^^^^
20:49:22 webpack.1 |    ╵
20:49:22 webpack.1 |     webpack/assets/javascripts/dashboard/gridster.scss 17:29  @import
20:49:22 webpack.1 |     webpack/assets/javascripts/dashboard/index.scss 1:9       root stylesheet
```

evgeni commented:
/packit build

MariaAga commented:
@jeremylenz 
> By the way, I see a lot of deprecation warnings like this. Are those being fixed?

Those are from pf3, and will go away once we stop using pf3.

evgeni commented:
The RPM build fails with:

```
+ export NODE_ENV=production
+ NODE_ENV=production
+ webpack --bail --config config/webpack.config.js
assets by status 3.32 MiB [cached] 3 assets
runtime modules 15.9 KiB 11 modules
modules by path ./webpack/assets/javascripts/react_app/ 3.38 MiB 587 modules
modules by path ../../../../usr/lib/node_modules/ 730 KiB 44 modules
modules by path ./webpack/assets/javascripts/*.js 81.3 KiB 23 modules
modules by path ./webpack/assets/javascripts/services/ 27.7 KiB 7 modules
modules by path ./webpack/assets/javascripts/compute_resource/*.js 19.4 KiB 5 modules
modules by path ./webpack/assets/javascripts/dashboard/ 14.3 KiB 3 modules
modules by path ./node_modules/ 572 bytes
  ./node_modules/react-intl/locale-data/ sync \/(ca%7Ccs%7Cde%7Cen%7Ces%7Cfr%7Cit%7Cja%7Cka%7Cko%7Cpl%7Cpt%7Cru%7Czh)$ 286 bytes [built] [code generated]
  ./node_modules/intl/locale-data/jsonp/ sync \/(ca%7Ccs%7Cde%7Cen%7Ces%7Cfr%7Cit%7Cja%7Cka%7Cko%7Cpl%7Cpt%7Cru%7Czh)$ 286 bytes [built] [code generated]
modules by path ./webpack/assets/javascripts/hosts/*.js 11 KiB
  ./webpack/assets/javascripts/hosts/tableCheckboxes.js 10.6 KiB [built] [code generated]
  ./webpack/assets/javascripts/hosts/HostsSessionStorage.js 490 bytes [built] [code generated]
+ 54 modules

ERROR in ./webpack/assets/javascripts/dashboard/index.scss (../../../../usr/lib/node_modules/css-loader/dist/cjs.js!../../../../usr/lib/node_modules/sass-loader/dist/cjs.js!./webpack/assets/javascripts/dashboard/index.scss)
Module Error (from ../../../../usr/lib/node_modules/sass-loader/dist/cjs.js):
Node Sass version 8.0.0 is incompatible with ^4.0.0 || ^5.0.0 || ^6.0.0.
 @ ./webpack/assets/javascripts/dashboard/index.scss 2:26-188
 @ ./webpack/assets/javascripts/dashboard/index.js 17:0-23
 @ ./webpack/assets/javascripts/bundle.js 25:40-62
```

Do I read https://github.com/webpack-contrib/sass-loader/commit/e5581b75e84879f27e221d67caa8507897e7051d correctly that we need sass-loader 13.2.0+ and not 10.2.0 like in the `package.json` here?

evgeni commented:
Or 10.4.1+: https://github.com/webpack-contrib/sass-loader/commits/v10

evgeni commented:
/packit build

evgeni commented:
FTR, an RPM build with sass-loader 10.4.1 succeeded.

MariaAga commented:
Looks like  sass-loader 13.2.0 is also fine for us so I updated to that

evgeni commented:
/packit build

evgeni commented:
It passed with 13.3.2 with rpm too, nice

evgeni commented on config/webpack.config.js:
```diff
@@ -242,46 +144,175 @@ module.exports = env => {
-    config.devtool = 'inline-source-map';
-  }
+  config.output = {
+    path: path.join(__dirname, '..', 'public', 'webpack', pluginName),
```
this used to be
```
    var outputPath = path.join(
      plugins['plugins'][env.pluginName]['root'],
      'public',
      'webpack'
    );
```
and I think the current notation breaks RPM builds of plugins.

evgeni commented on config/webpack.config.js:
```diff
@@ -242,46 +144,175 @@ module.exports = env => {
-    config.devtool = 'inline-source-map';
-  }
+  config.output = {
+    path: path.join(pluginRoot, 'public', 'webpack', pluginName),
```
using pluginRoot here made RPM builds work, but I guess it breaks devel setups?

evgeni replied:
seems this reliably hangs integration tests on GHA, nice

MariaAga replied:
We'll need to also make it so it will be served as I get `No route matches [GET] "/webpack/foreman_ansible/foreman_ansible_remoteEntry.js"` 
and add a git ignore to plugins :sweat_smile:

evgeni replied:
huh, seems the CI failure is unrelated to my changes, as it also fails on plain `develop`: https://github.com/theforeman/foreman/actions/runs/7213337369/job/19653057456?pr=9955 :exploding_head:

evgeni replied:
the issue is foreman-js 12.2.1 :see_no_evil:

evgeni replied:
> We'll need to also make it so it will be served

This is fixed :)

evgeni commented:
After the following changes, things seem to work fine also for RPM builds:
* put the webpack result of plugins in different output folders, depending on prod/dev mode -- they are served differently and we had this differentiation with webpack3 too
* back out `--analyze` out of the `Procfile` as suggested by Maria in chat
* pin foreman-js to 12.2.0 as 12.2.1 seems to break things (and I didn't want to go hunt that one down)

But there is one thing I'd like to discuss. You've started calling `webpack` via `npx` (and thus require the additional `webpack-cli` package), but it seems our usage is inconsistent.

In packaging, when building Foreman, [we call `webpack` directly](https://github.com/theforeman/foreman-packaging/blob/3bb99a3adbcd761d15e652ca274a9a4685b5766e/packages/foreman/foreman/foreman.spec#L564):
```
webpack --bail --config config/webpack.config.js
```
While the rake tasks used `node --max_old_space_size=2048 #{webpack_bin} --config #{config_file} --bail` before this PR, and are using `npx … webpack …` after it. (interestingly, in the Debian package we *do* call `rake webpack:compile`, not `webpack` directly)

For plugins, the rake tasks used `node --max_old_space_size=2048 #{webpack_bin} --config #{config_file} --bail --env.pluginName=#{@plugin.id}` and now use `npx … webpack …` and packaging just calls these.

@jeremylenz has described `npx` as "first hunt in node_modules, then if I don't find something, npm install it", which seems like something you'd use in an interactive shell, but not in Rake tasks or packaging, as there you have installed all required dependencies before? On the other hand, it doesn't hurt much there either, as the deps are already installed and it will just find the binary on disk.

I guess calling `npx webpack` has the benefit that we do not have to care where and how webpack was installed, and `npx` will just find it, so the question is: should we switch to using `npx webpack` everywhere (= change RPM packaging to use `rake webpack:compile`) for consistency?

evgeni commented on lib/tasks/jenkins.rake:
```diff
@@ -3,7 +3,7 @@ begin
   namespace :jenkins do
     task :unit => ['jenkins:setup:minitest', 'rake:test:units', 'rake:test:functionals', 'rake:test:graphql']
-    task :integration => ['webpack:compile', 'jenkins:setup:minitest', 'rake:test:integration']
+    task :integration => ['webpack:compile', 'assets:precompile', 'jenkins:setup:minitest', 'rake:test:integration']
```
why was this needed? integration tests on GHA don't use `jenkins:integration` but call `webpack:compile test:integration` directly and still pass fine

evgeni commented on script/foreman-start-dev:
```diff
@@ -1,3 +1,3 @@
 #!/bin/sh
-./node_modules/.bin/webpack-dev-server-without-h2 --config config/webpack.config.js --host "::" $WEBPACK_OPTS &
+npx webpack --config config/webpack.config.js --watch $WEBPACK_OPTS &
 ./bin/rails server -b \[::\] "$@"
```
a tad unrelated, but isn't this script doing essentially the same as `bundle exec foreman start`, and could be just dropped?

Ron-Lavi commented on app/assets/javascripts/application.js:
```diff
@@ -7,9 +7,22 @@
+    $(document).trigger('ContentLoad');
+  }
+  else {
+  $(document).on('loadJS', function() {
```
not sure if it's related to this change, but when moving from legacy UI page to a React page, on every API `get` request I see the following error:

![image](https://github.com/theforeman/foreman/assets/26363699/d3eec3d4-306a-4fb4-b06d-646fe2028fc1)


I believe that `Axios` is using the DOM to store some of the request properties, and when trying to access an element it fails to find it, could be some race condition.

overall it doesn't seem to affect the pages functionality.

MariaAga replied:
I cant reproduce this, which pages are you moving from/to?

Ron-Lavi replied:
got fixed for me in the latest commit and @theforeman/vendor v12.2.3

Ron-Lavi commented on webpack/assets/javascripts/react_app/common/AwaitedMount.js:
```diff
@@ -0,0 +1,60 @@
+import PropTypes from 'prop-types';
+import { useState, useEffect } from 'react';
+import store from '../redux';
+import componentRegistry from '../components/componentRegistry';
+
+// Mounts a component after all plugins have been imported to make sure that all plugins are available to the component
+export const AwaitedMount = ({ component, data, flattenData }) => {
```
there's a warning on every page about `Component name already taken:` 
from the stack trace it's originated from this file.

![image](https://github.com/theforeman/foreman/assets/26363699/b1ae6e42-dde2-4166-a1e0-3ea5b24771b7)

![image](https://github.com/theforeman/foreman/assets/26363699/5900ce80-7bc4-44d0-9843-c0ade0f5ac20)

evgeni replied:
Was this compiled with foreman-js 12.2.2? If so, that's the real culprit, not this line.
12.2.3 should fix everything.

Ron-Lavi replied:
should we drop the version pin to v12.2.0 now? https://github.com/theforeman/foreman/pull/9834/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R24 ?

going to try with v12.2.3

evgeni replied:
12.2.3 should be fine. only .1 and .2 are broken.

I'll drop the pin, good catch

Ron-Lavi replied:
yeah looks like this is fixed in the latest commit here and using v12.2.3

Ron-Lavi replied:
apologies, there was an issue that plugins weren't compiling well using `bundle exec rake webpack:compile`
it worked with `npx webpack  --config config/webpack.config.js`

and now the issue is visible again

Ron-Lavi commented on app/views/layouts/base.html.erb:
```diff
@@ -33,18 +30,20 @@
     <%= javascript_include_tag "locale/#{FastGettext.locale}/app" %>
     <%= locale_js_tags %>
-    
     <%= yield(:head) %>
```
just before the end of the <head> tags we are getting now many <style> tags which I think comes from this yield:

![image](https://github.com/theforeman/foreman/assets/26363699/36e5585f-b4d1-49de-9f6c-54ac22e18778)

as of now, I don't see an impact on existing pages.

Ron-Lavi reviewed (commented):
Nice work @MariaAga !! thanks for working hard on getting this in.

Something that we should pay attention to regarding dev environment compilation, 
I made a change in a string and after 8 seconds the webpack compiled,
later on regular refresh the string didn't change, only after hard refresh it did.

after that I changed the string again and even when webpack got compiled successfully the page with the string remained the same.

is there any caching mechanism we are using?

Thanks

Ron-Lavi reviewed (commented):
@MariaAga  Can you check if clicking on the kebab toggle in the host details page fails for you as well?

I got some error regarding `onKebabToggle` being `undefined`, I wonder if there is some issue with the React Context provider.

verified and it works on the develop branch.

MariaAga commented:
Also saw this issue now, that I will fix
![image](https://github.com/theforeman/foreman/assets/30431079/cf82b632-dcf9-4001-8e9e-65e24bb4b3f1)

ekohl commented:
I played used this while working on https://github.com/theforeman/foreman_monitoring/pull/94 and as a developer there it appeared to work as I expected.

Ron-Lavi commented:
> @MariaAga Can you check if clicking on the kebab toggle in the host details page fails for you as well?
> 
> I got some error regarding `onKebabToggle` being `undefined`, I wonder if there is some issue with the React Context provider.
> 
> verified and it works on the develop branch.

Update: Works for me on the latest commit with @theforeman/vendor v12.2.3

MariaAga commented:
`I got some error regarding onKebabToggle being undefined, I wonder if there is some issue with the React Context provider.` 
This is still happening with bootdisk plugin I think. I have a solution for it, will push when I'll resolve the rest of the issues (400 style tags, and deprecation text in the UI)

Ron-Lavi commented:
> `I got some error regarding onKebabToggle being undefined, I wonder if there is some issue with the React Context provider.` This is still happening with bootdisk plugin I think. I have a solution for it, will push when I'll resolve the rest of the issues (400 style tags, and deprecation text in the UI)

apologies, there was an issue that plugins weren't compiling well using `bundle exec rake webpack:compile`
it worked with `npx webpack  --config config/webpack.config.js`

and now the issue with clicking the host details kebab is happening again, my plugins are:
![image](https://github.com/theforeman/foreman/assets/26363699/2dba773a-d187-46b2-95ae-78a6b98fd24c)

MariaAga commented:
* `bundle exec rake webpack:compile`  is defaulting to `NODE_ENV=production` I'll remove/edit the suggestion to use it.
* Instead of adding context files (like `Templates/CardItem/CardTemplate/index.js` and `HostDetails/ActionsBar/index.js`) to `ModuleFederationPlugin` `shared` field, I have wrapped the contexts with `forceSingleton` which resolved the issues in host details page.
 * Changed `style-loader` to use `singletonStyleTag` -  https://webpack.js.org/loaders/style-loader/#singletonstyletag
 Added an id=plugin so we can track which plugin css rules come from as this option disables source maps.
![image](https://github.com/theforeman/foreman/assets/30431079/2269cdd4-d50c-4d79-96f8-8243aff6e0ee)

ekohl commented on Procfile:
```diff
@@ -2,4 +2,7 @@
 # you can use WEBPACK_OPTS to customize webpack server, e.g. 'WEBPACK_OPTS='--https --key /path/to/key --cert /path/to/cert.pem --cacert /path/to/cacert.pem' foreman start '
-webpack: [ -n "$NODE_ENV" ] && ./node_modules/.bin/webpack-dev-server-without-h2 --config config/webpack.config.js $WEBPACK_OPTS || env NODE_ENV=development ./node_modules/.bin/webpack-dev-server-without-h2 --config config/webpack.config.js $WEBPACK_OPTS
+webpack: [ -n "$NODE_ENV" ] && npx webpack --config config/webpack.config.js --watch $WEBPACK_OPTS || env NODE_ENV=development npx webpack --config config/webpack.config.js --watch
+#TODO readme/forklift to change/remove --key to --server-options-key
+# --public -> --client-web-socket-url and see why it doesnt work
+# --https --> --server-type
```
This is https://github.com/theforeman/puppet-katello_devel/blob/596745918fa758f209dbffb81d673c6c011186c9/templates/env.erb#L4. It's not exactly clear to me what needs to change. Can you submit a PR?

MariaAga replied:
Since we wont be using webpack as a server, we need to remove those.
I'm not sure when will be a good time to remove it in case people (like me) use forklift for older foreman versions.
For now I'm filtering these options out with `FILTERED_WEBPACK_OPTS` but if users somehow use these options in webpack (for example, running `npx webpack --config config/webpack.config.js --https`) webpack will not run.

ekohl replied:
I think that's an acceptable solution.

ekohl commented on app/helpers/layout_helper.rb:
```diff
@@ -98,6 +98,29 @@ def javascript(*args)
+  def javascript_include_tag(*params, **kwargs)
+    # Workaround for overriding javascript load with webpack_asset_paths, should be removed when webpack_asset_paths is removed
+    if kwargs[:source] == "webpack_asset_paths"
+      kwargs[:webpacked]
```
There's no `super` call here, so it does nothing. Is that intentional, or did you mean to unconditionally call `super(...)`?

MariaAga replied:
we dont want to call super cause it returns a script tag, 
and the webpacked key contains for example: `webpacked=>"<script>\n//<![CDATA[\nwindow.tfm.tools.loadPluginModule('/webpack/foreman_remote_execution','foreman_remote_execution','./index');\n//]]>\n</script>"`
which we declare  in: ` webpacked: webpacked_plugins_js_for(plugin_name.to_sym)`

MariaAga commented on app/helpers/reactjs_helper.rb:
```diff
@@ -24,7 +21,25 @@ def webpacked_plugins_with_global_js
 
   def webpacked_plugins_css_for(*plugin_names)
-    css_tags_for(select_requested_plugins(plugin_names)).join.html_safe
+    Foreman::Deprecation.deprecation_warning('3.12', '`webpacked_plugins_css_for` is deprecated, plugin css is already loaded.')
```
small todo: return nil after this line as sometimes the deprecation warning is shown in the dom above the foreman header

evgeni commented on package.json:
```diff
@@ -20,56 +20,54 @@
   "dependencies": {
-    "@theforeman/vendor": "^12.0.1",
+    "@module-federation/utilities": "^1.7.0",
+    "@theforeman/vendor": "12.2.0",
```
You can drop the commit that pins that to .0

ekohl commented:
@MariaAga and all reviewers: I'm proposing we set a date to merge this and deal with any fallout after. We should announce this date on Discourse and other places so people are aware of it. I'm proposing Monday 29 or Tuesday 30. That gives us a full week to iterate on it and prepare, while also giving us a week before FOSDEM + cfgmgmtcamp. Looking at the [3.10 schedule](https://community.theforeman.org/t/foreman-3-10-schedule-and-planning/36482) we'd have about 2 weeks until stabilization week to deal with any fallout, or roll back if it's really too bad. How does that sound?

evgeni commented on .packit.yaml:
```diff
@@ -18,7 +18,7 @@ downstream_package_name: foreman
 
 actions:
   post-upstream-clone:
-    - "wget https://raw.githubusercontent.com/theforeman/foreman-packaging/rpm/develop/packages/foreman/foreman/foreman.spec -O foreman.spec"
```
we should not forget to drop the `[HACK] use evgeni's temporary repo with webpack5 builds` commit before merging (we can also do so now, the only downside would be failing packit builds, but that's IMHO fine)

evgeni commented:
> I'm proposing Monday 29 or Tuesday 30

Can we aim for Friday the 26th instead? That would allow us to come in fresh on Monday and see what the pipelines and everyone has to say about the result.

ekohl commented:
I'm not a fan of merging something potentially breaking on Friday, but I can see arguments for letting pipelines run. I can't speak for everyone, but my Monday is also pretty full with meetings which why I also said Tuesday. Friday would be OK with me. If people get stuck with things on the weekend, they can submit PRs and we can merge them while waiting on pipelines during meetings.

MariaAga commented on config/webpack.config.js:
```diff
@@ -219,8 +219,13 @@ const pluginConfig = function(plugin) {
   });
 
+  if (config.mode == 'production') {
+    var outputPath = path.join(pluginRoot, 'public', 'webpack', pluginName);
```
@evgeni Why was this change made? should we add a .gitignore to all plugins for it?

evgeni replied:
When we compile production assets (= during package building), we don't have write access to the Foreman folder, so we place the assets into the plugin directory (and then create a symlink when the package is installed).
Without this change, RPM builds of plugins fail, as they don't find the assets where they expect them to be.

This matches this old code:
https://github.com/theforeman/foreman/blob/066dd49a9ec121038bd93ac7d03f28ea0fc4b69c/config/webpack.config.js#L88-L92
introduced in https://github.com/theforeman/foreman/commit/b0ee9fd635191cf890fa5f63e26f7830c829f552

As this is a previously used path, and only used in production (where we don't have Git checkouts), I don't think this deserves any `.gitignore` additions.

MariaAga committed:
Fixes #37102 - webpack 5

ekohl reviewed (approved):
Let's break some pipelines and development environments by merging this :)

As a reminder, https://community.theforeman.org/t/webpack-5-merge-this-friday-2024-01-26/36581 recommends removing your `node_modules` directory entirely and run `npm install` fresh.
